### PR TITLE
Add QRCodeScanner widget

### PR DESCRIPTION
**Pull request by vinothvino42**
Ticket: https://github.com/EnsembleUI/ensemble/issues/1254

Other PRs
ensemble - https://github.com/EnsembleUI/ensemble/pull/1260
starter - https://github.com/EnsembleUI/starter/pull/41

**Sample EDL**:

```yaml
View:
  header:
    title: QR Code Scanner

  body:
    Column:
      styles: { gap: 20, margin: 10, crossAxis: center, mainAxisSize: min }
      children:
        - Column:
            styles: { gap: 10, crossAxis: center }
            children:
              - Text:
                  id: qrFormatId
                  text: "QR Code Format: "
              - Text:
                  id: qrDataId
                  text: "QR Code Data: "

              - Divider:

              - Flow:
                  styles: { gap: 20 }
                  children:
                    - Text:
                        text: IsFlashOn - ${qrCodeScannerId.isFlashOn}
                    - Text:
                        text: hasFlash - ${qrCodeScannerId.hasFlash}
                    - Text:
                        text: hasBackCamera - ${qrCodeScannerId.hasBackCamera}
                    - Text:
                        text: hasFrontCamera - ${qrCodeScannerId.hasFrontCamera}

        - QRCodeScanner:
            id: qrCodeScannerId
            styles:
              expanded: true
              width: ${device.width}

              # overlayColor: grey
              overlayMargin: 20
              cutOutBorderColor: orange
              cutOutBorderLength: 50
              cutOutBorderWidth: 12
              cutOutBorderRadius: 20
              cutOutHeight: 250
              cutOutWidth: 250
            formatsAllowed:
              - qrcode
              - rss14
            initialCamera: back
            onInitialized: |
              //@code
              console.log("View Created");
            onPermissionSet: |
              //@code
              console.log("Permission Set: " + event.data.status);
            onReceived: |
              //@code
              qrFormatId.text = "QR Code Format: " + event.data.format;
              qrDataId.text = "QR Code Data: " + event.data.data;

              // Printing it to console
              console.log("Format: "+event.data.format);  // qrcode, rss14, upcA, aztec, etc
              console.log("Code: "+event.data.data);
              console.log("Raw Bytes: "+event.data.rawBytes); // Only supported for Android & iOS

        - Flow:
            styles: { margin: 20, gap: 10, lineGap: 10, mainAxis: center }
            children:
              - Button:
                  label: Flip Camera
                  onTap: |
                    //@code
                    qrCodeScannerId.flipCamera();

              - Button:
                  label: Toggle Flash
                  onTap: |
                    //@code
                    qrCodeScannerId.toggleFlash();

              - Button:
                  label: Pause Camera
                  onTap: |
                    //@code
                    qrCodeScannerId.pauseCamera();

              - Button:
                  label: Resume Camera
                  onTap: |
                    //@code
                    qrCodeScannerId.resumeCamera();

```

### CODE REVIEW

**Comment by kmahmood74** [comment]
@vinothvino42 can you post a video of this widget working in ensemble?

File: lib/qr_code_scanner.dart [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+    }
+  }
+
+  @override
+  void flipCamera() {
+    qrController?.flipCamera();
+  }
+
+  // TODO:
+  void getSystemFeatures() async {
```
Comment (kmahmood74) [inline lib/qr_code_scanner.dart]: @vinothvino42 what is this for anyway?

**Review by kmahmood74 — COMMENTED** [review]
there are a lot of options here, have you tested all of them? 
will users be able to open the scanner as a full screen camera?
will it close automatically and send onReceive as soon as a QR code is scanned?

File: lib/qr_code_scanner.dart (hunk 1 above)
Comment (kmahmood74) [reply lib/qr_code_scanner.dart]: please also upload a video of this working

**Comment by vinothvino42** [comment]
> there are a lot of options here, have you tested all of them? will users be able to open the scanner as a full-screen camera? will it close automatically and send onReceive as soon as a QR code is scanned?

1. Yes, I tested all the options
2. In this case, the user has to create a new screen with full device height and width passed to the QRCodeScanner widget. 
3. We need to manually call `navigateBack` action from `onReceived` callback once you get the data

File: lib/qr_code_scanner.dart (hunk 1 above)
Comment (vinothvino42) [reply lib/qr_code_scanner.dart]: This is to get the device capabilities for scanning a qrcode, whether its having back and front camera, flash, etc
https://pub.dev/documentation/qr_code_scanner/latest/qr_code_scanner/SystemFeatures-class.html

Comment (vinothvino42) [reply lib/qr_code_scanner.dart]: > please also upload a video of this working

Sure, I'll upload and send you the link now

**Comment by vinothvino42** [comment]
@kmahmood74

I added a stub widget in the ensemble repo and updated the starter repo to use the `QRCodeScanner` Widget. I updated the first comment with other PRs and EDL. https://github.com/EnsembleUI/ensemble_camera/pull/11#issue-2147186424

File: lib/qr_code_scanner.dart [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+    await qrController?.toggleFlash();
+    widget.controller.isFlashOn = !widget.controller.isFlashOn;
+    dispatchValue('isFlashOn', widget.controller.isFlashOn);
+  }
+
+  void dispatchValue(String key, dynamic value) {
+    ScopeManager? scopeManager = ScreenController().getScopeManager(context);
+    if (scopeManager == null || widget.controller.id == null) return;
+
+    scopeManager.dataContext.addDataContextById(key, value);
```
Comment (kmahmood74) [inline lib/qr_code_scanner.dart]: why are we adding these directly to the context here?

File: lib/qr_code_scanner.dart [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+  }
+
+  Future<void> _getSystemFeatures() async {
+    try {
+      final systemFeatures = await qrController?.getSystemFeatures();
+      if (systemFeatures == null) return;
+      widget.controller.hasFlash = systemFeatures.hasFlash;
+      widget.controller.hasBackCamera = systemFeatures.hasBackCamera;
+      widget.controller.hasFrontCamera = systemFeatures.hasFrontCamera;
+      dispatchValue('hasFlash', widget.controller.hasFlash);
```
Comment (kmahmood74) [inline lib/qr_code_scanner.dart]: these are system wide capabilities, they can't just be hidden under the qrcode scanner. what's the point of doing this here?

File: lib/qr_code_scanner.dart [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+    dispatchValue('isFlashOn', widget.controller.isFlashOn);
+  }
+
+  void dispatchValue(String key, dynamic value) {
+    ScopeManager? scopeManager = ScreenController().getScopeManager(context);
+    if (scopeManager == null || widget.controller.id == null) return;
+
+    scopeManager.dataContext.addDataContextById(key, value);
+    final data = scopeManager.dataContext.getContextById(widget.controller.id!);
+    scopeManager.dispatch(
```
Comment (kmahmood74) [inline lib/qr_code_scanner.dart]: model hasn't changed, why are you firing this event?

File: lib/qr_code_scanner.dart [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+          widget.controller.onReceived!,
+          event: EnsembleEvent(widget.controller, data: data),
+        );
+      }
+    });
+  }
+
+  void _onPermissionSet(BuildContext context, QRViewController ctrl, bool p) {
+    if (!p) {
+      ScaffoldMessenger.of(context).showSnackBar(
```
Comment (kmahmood74) [inline lib/qr_code_scanner.dart]: developer should be able to call any action on this, we can't just show a snackbar ourserlves

File: lib/qr_code_scanner.dart [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+        .toList();
+    return allFormats;
+  }
+
+  @override
+  Map<String, Function> getters() {
+    return {
+      'formatsAllowed': () => formatsAllowed,
+      'mode': () => mode.name,
+      'hasBackCamera': () => hasBackCamera,
```
Comment (kmahmood74) [inline lib/qr_code_scanner.dart]: all these properties are being set inside the build method after a significant delay. So what would be the value if the user calls these before they are available?

File: lib/qr_code_scanner.dart [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+      'hasFrontCamera': () => hasFrontCamera,
+      'hasFlash': () => hasFlash,
+      'isFlashOn': () => isFlashOn,
+    };
+  }
+
+  @override
+  Map<String, Function> methods() => Map<String, Function>.from(super.methods())
+    ..addAll({
```
Comment (kmahmood74) [inline lib/qr_code_scanner.dart]: you don't do allAll from super, you just specify your widget's methods

File: lib/qr_code_scanner.dart [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+    widget.controller.isFlashOn = !widget.controller.isFlashOn;
+    dispatchValue('isFlashOn', widget.controller.isFlashOn);
+  }
+
+  void dispatchValue(String key, dynamic value) {
+    ScopeManager? scopeManager = ScreenController().getScopeManager(context);
+    if (scopeManager == null || widget.controller.id == null) return;
+
+    scopeManager.dataContext.addDataContextById(key, value);
+    final data = scopeManager.dataContext.getContextById(widget.controller.id!);
```
Comment (kmahmood74) [inline lib/qr_code_scanner.dart]: widget.controller.id? where is that and what if it is not set?

File: lib/qr_code_scanner.dart [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+  Future<void> _getSystemFeatures() async {
+    try {
+      final systemFeatures = await qrController?.getSystemFeatures();
+      if (systemFeatures == null) return;
+      widget.controller.hasFlash = systemFeatures.hasFlash;
+      widget.controller.hasBackCamera = systemFeatures.hasBackCamera;
+      widget.controller.hasFrontCamera = systemFeatures.hasFrontCamera;
+      dispatchValue('hasFlash', widget.controller.hasFlash);
+      dispatchValue('hasBackCamera', widget.controller.hasBackCamera);
+      dispatchValue('hasFrontCamera', widget.controller.hasFrontCamera);
```
Comment (kmahmood74) [inline lib/qr_code_scanner.dart]: why are we dispatching these values?

**Review by kmahmood74 — CHANGES_REQUESTED** [review]
@vinothvino42 this is unnecessarily complicated and error prone. All a dev would need to know is this - 

1. Does the device have camera? - should have a clear error handler to call so dev can decide what to do
2. Was the camera enabled? - should have a clear error handler to call so dev can decide what to do
3. Camera should come with flash, flip view and all those things. I don't need to change those capabilities
4. and then of course the qr code and it's properties
5. ability to open the camera, pause and close it

I think this **also** needs to be an Action just like the openCamera action

File: lib/qr_code_scanner.dart (hunk 2 above)
Comment (vinothvino42) [reply lib/qr_code_scanner.dart]: All these properties (hasFlash, hasBackCamera, hasFrontCamera, isFlashOn) are only a getter. There is a method in the package called getSystemFeatures, I guess it's not need to expose it to the EDL this method. So just calling this method internally and exposing only the getter property to EDL

File: lib/qr_code_scanner.dart (hunk 3 above)
Comment (vinothvino42) [reply lib/qr_code_scanner.dart]: Same - https://github.com/EnsembleUI/ensemble_camera/pull/11#discussion_r1500223489

File: lib/qr_code_scanner.dart (hunk 4 above)
Comment (vinothvino42) [reply lib/qr_code_scanner.dart]: To update the getter property

File: lib/qr_code_scanner.dart (hunk 5 above)
Comment (vinothvino42) [reply lib/qr_code_scanner.dart]: Okay I'll expose `onPermissionSet` callback with the boolean value in the ensemble event data (`event.data.status`) to the EDL.

File: lib/qr_code_scanner.dart (hunk 6 above)
Comment (vinothvino42) [reply lib/qr_code_scanner.dart]: The value will be default (false) for all those properties

File: lib/qr_code_scanner.dart (hunk 7 above)
Comment (vinothvino42) [reply lib/qr_code_scanner.dart]: Okay I'll update it

File: lib/qr_code_scanner.dart (hunk 8 above)
Comment (vinothvino42) [reply lib/qr_code_scanner.dart]: The `id` of the `EnsembleWidgetController`

File: lib/qr_code_scanner.dart (hunk 9 above)
Comment (vinothvino42) [reply lib/qr_code_scanner.dart]: Same - https://github.com/EnsembleUI/ensemble_camera/pull/11#discussion_r1500223489

File: lib/qr_code_scanner.dart (hunk 6 above)
Comment (kmahmood74) [reply lib/qr_code_scanner.dart]: I know but that's wrong, how would the user know to call this after a delay?

File: lib/qr_code_scanner.dart (hunk 2 above)
Comment (kmahmood74) [reply lib/qr_code_scanner.dart]: remove this please. If we are checking system-wide capabilities, they shouldn't be hidden in a qr code widget. They should be available as `device.capabilities` everywhere whether this app uses qr code widget or not. Open a separate ticket to expose these capabilities at the device level.

File: lib/qr_code_scanner.dart (hunk 8 above)
Comment (kmahmood74) [reply lib/qr_code_scanner.dart]: that id is the id of the field if it is set and it is optional!

File: lib/qr_code_scanner.dart (hunk 9 above)
Comment (kmahmood74) [reply lib/qr_code_scanner.dart]: there is no reason to dispatch value, all you need to do is to set the widget.controller.xxxxx = <new value>

File: lib/qr_code_scanner.dart [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+  }
+
+  void dispatchValue(String key, dynamic value) {
+    ScopeManager? scopeManager = ScreenController().getScopeManager(context);
+    if (scopeManager == null || widget.controller.id == null) return;
+
+    scopeManager.dataContext.addDataContextById(key, value);
+    final data = scopeManager.dataContext.getContextById(widget.controller.id!);
+    scopeManager.dispatch(
+        ModelChangeEvent(SimpleBindingSource(widget.controller.id!), data));
```
Comment (kmahmood74) [inline lib/qr_code_scanner.dart]: what if id is null? it is an optional field!

File: lib/qr_code_scanner.dart (hunk 4 above)
Comment (kmahmood74) [reply lib/qr_code_scanner.dart]: no, all you need to do it is to call `setProperty` method on your controller and the framework will take care of the rest. Call setProperty to set he properties and don't set them the way you are setting them. No modelchange event needed. Look at how other widgets work

File: lib/qr_code_scanner.dart [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+
+  @override
+  void resumeCamera() {
+    qrController?.resumeCamera();
+  }
+
+  @override
+  void toggleFlash() async {
+    await qrController?.toggleFlash();
+    widget.controller.isFlashOn = !widget.controller.isFlashOn;
```
Comment (kmahmood74) [inline lib/qr_code_scanner.dart:229]: use setProperty and remove dispatchValue

**Review by kmahmood74 — CHANGES_REQUESTED** [review]
this is not right, 

1. remove all the dispatchValue stuff
2. simplify this significantly
3. show me how one may open the scanner camera in full-screen

File: lib/qr_code_scanner.dart (hunk 2 above)
Comment (vinothvino42) [reply lib/qr_code_scanner.dart]: Okay I'll create a new ticket for it and will remove these system features from this widget

**Review by kmahmood74 — APPROVED** [review]
what's onPermissionSet?

**Comment by kmahmood74** [comment]
Please create a kitchen sink example and link it here
please create documentation and link it here

**Comment by kmahmood74** [comment]
sorry link them to the ticket, not on the PR

**Comment by vinothvino42** [comment]
> what's onPermissionSet?
It's the camera permission

**Comment by vinothvino42** [comment]
> Please create a kitchen sink example and link it here please create documentation and link it here

https://github.com/EnsembleUI/ensemble/issues/1254#issuecomment-1969181770

**Pull request merged**